### DISCUSSION TITLE
아이템 24. 멤버 클래스는 되도록 static으로 만들라

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#48] [3. 과제] [answered]
> 📝 구성
- [1. 중첩 클래스 정의 ✨ ](#1.-중첩-클래스(nested-class)-정의)
- [2. 중첩 클래스 종류 📱 ](#2.-중첩-클래스-종류)
- [3. 중첩 클래스의 사용 당위성 📚](#3------------------)
- [4. 회고 🧹](#4------)

<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
--- 

# 1. 중첩 클래스(nested class) 정의 ✨
- 중첩 클래스란?
- 클래스 내부에 선언한 클래스를 말한다.
  - 외부 클래스와 내부 클래스간의 멤버 접근을 서로 쉽게 할 수 있다.
  - 외부(_여기서 말하는 외부는 외부 클래스가 아님_)에는 불 필요한 클래스간의 관계를 감춰 코드 복잡성이 줄어든다. - a.k.a 캡슐화
```java
class Outer {
    // Some Fields...
    class Inner {
        // Some Fields...
    }
}
```
---

# 2. 중첩 클래스 종류 📱
### 2-1. 정적 멤버 클래스 - static Class
![image](https://user-images.githubusercontent.com/53285909/211316959-7244bdb7-a9ab-4cd0-b51d-d212430c453c.png)

- 정적 멤버 클래스는 다른 클래스 안에 선언되고, 바깥 클래스의 private 멤버에도 접근 할 수 있다는 점을 제외한다면 일반 클래스와 같다.
```java
class RecommendService {
    public String getRecommendId(String new_id) {
        String id = new IdMaker(new_id)
                .replaceToLowerCase()
                .replacePattern()
                .replaceGreaterThan16()
                .getId();

        return id;
    }

    private static class IdMaker {
        private String id;

        IdMaker(String id) {
            this.id = id;
        }

        private IdMaker replaceToLowerCase(){
            id = id.toLowerCase();
            return this;
        }

        private IdMaker replacePattern(){
            id = id.replaceAll("[^a-z0-9]", "");
            return this;
        }

        private IdMaker replaceGreaterThan16(){
            if (id.length() > 15) {
                id = id.substring(0, 15);
            }
            return this;
        }

        private String getId(){
            return id;
        }
    }
}
```
<details>
<summary>책 예제 코드 보기</summary>
<div markdown="1">

```java
public class Calculator {
    public enum Operation { // 열거 타입도 암시적 static 이다.
        PLUS, MINUS, MULTIPLE, SUBTRACT
    }
}
```

</div>
</details>

- 정적 멤버 클래스는 다른 정적 멤버와 똑같은 규칙을 적용받는다.
- 위의 예제에서 보면 알 수 있듯이 private으로 `InnerFoo`를 선언하면 `OuterFoo` 에서만 접근 가능하다.
> Q. `정적 멤버 클래스`는 언제 사용할까? 🤔
> > A. `외부 클래스`의 _인스턴스 생성없이_ `내부 클래스`에 접근하기 위한 용도이다.
그러므로 내부 클래스를 외부에 노출 시키려는 용도로 사용하고 싶을 경우에는 내부 클래스에 static을 붙이는 게 외부 클래스의 인스턴스 생성 없이 사용 가능하므로 사용하기에 유용하다.

---

### 2-2. 비정적 멤버 클래스 - non static Class
- 비정적 멤버 클래스의 인스턴스와 바깥 인스턴스 사이의 관계는 멤버 클래스가 인스턴스화 될 때 확립되며, 더 이상 변경할 수 없다.

<details>
<summary>코드 보기</summary>
<div markdown="1">

```java
class Foo {

  void bar() {
  }

  class NestedFoo {

    void bar() {
      Foo.this.bar();
    }
  }
}

class Main {

  public static void main(String[] args) {
    NestedFoo nestedFoo = new Foo().new NestedFoo();
    nestedFoo.bar();
  }
}
```

</div>
</details>

- 비정적 멤버 클래스`NestedFoo`의 인스턴스는 바깥 클래스인 `Foo` 와 암묵적으로 연결된다.
- 이는 비정적 멤버 클래스의 인스턴스 메서드에 정규화된 `this`를 사용해 바깥 인스턴스의 메서드를 호출하거나 바깥 인스턴스의 참조를 가져올 수 있다.
  - 하지만 해당 행위는 비정적 멤버 클래스의 인스턴스 안에 만들어져 메모리 공간을 차지하며, 생성 시간도 더 느리다. 

> Q. `비정적 멤버 클래스`는 언제 사용할까? 🤔
>
> A. [어댑터 패턴](https://refactoring.guru/ko/design-patterns/adapter)에 이용된다.
>  즉 다른 타입의 인스턴스를 리턴할 때 이용된다.

<details>
<summary>HashMap에 사용중인 비정적 멤버 클래스 코드 보기</summary>
<div markdown="1">

```java
public class HashMap<K,V> extends AbstractMap<K,V>
    implements Map<K,V>, Cloneable, Serializable {
    ...
    public Set<K> keySet() {
        Set<K> ks = keySet;
        if (ks == null) {
            ks = new KeySet();
            keySet = ks;
        }
        return ks;
    }
    final class KeySet extends AbstractSet<K> {
        public final int size()                 { return size; }
        public final void clear()               { HashMap.this.clear(); }
        public final Iterator<K> iterator()     { return new KeyIterator(); }
        public final boolean contains(Object o) { return containsKey(o); }
        public final boolean remove(Object key) {
            return removeNode(hash(key), key, null, false, true) != null;
        }
        public final Spliterator<K> spliterator() {
            return new KeySpliterator<>(HashMap.this, 0, -1, 0, 0);
        }
        public final void forEach(Consumer<? super K> action) {
            Node<K,V>[] tab;
            if (action == null)
                throw new NullPointerException();
            if (size > 0 && (tab = table) != null) {
                int mc = modCount;
                for (Node<K,V> e : tab) {
                    for (; e != null; e = e.next)
                        action.accept(e.key);
                }
                if (modCount != mc)
                    throw new ConcurrentModificationException();
            }
        }
    }
}
```
</div>
</details>

다른 타입의 인스턴스를 리턴한다라는 말을 좀 더 풀어 보자면
- 어떤 클래스의 인스턴스를 감싸 마치 다른 클래스의 인스턴스처럼 보이게 하는 뷰로 사용 한다는 것이다.
- 위의 코드와 비슷하게 Set과 List 같은 다른 컬렉션 인터페이스 구현들에도 이용된다.

<details>
<summary>AbstractSet의 구현 코드</summary>
<div markdown="1">

```java
package effective;

import java.util.AbstractSet;
import java.util.Iterator;

public class MySet<E> extends AbstractSet<E> {

    @Override
    public Iterator<E> iterator() {
        return new MyIterator();
    }

    @Override
    public int size() {
        return 0;
    }
    
    private class MyIterator implements Iterator<E> {

        @Override
        public boolean hasNext() {
            return false;
        }

        @Override
        public E next() {
            return null;
        }
    }
}
```
</div>
</details>

- **멤버 클래스에서 바깥 인스턴스에 접근할 일이 없다면 무조건 `static`을 붙여서 정적 멤버 클래스로 만들자.**
  - 왜?
  - static의 생략은 바깥 인스턴스로의 숨은 외부 참조를 갖게 된다.
  - 숨은 외부 참조를 갖기 위해 시간과 공간이 소비된다.
  - ⭐️ 가비지 컬렉션이 바깥 클래스의 인스턴스를 수거하지 못하는 문제를 겪게 된다.
  - 이는 메모리 누수로 이어진다.
- 멤버 클래스가 공개된 클래스의 public이나 protected 멤버라면 정적이냐 아니냐는 두 배로 중요해진다. 멤버 클래스 역시 공개 API가 되니, 혹시라도 향후 릴리스에서 static을 붙이면 하위 호환성이 깨진다.

---

### 2-3. 익명 클래스

- 익명 클래스는 바깥 클래스의 멤버가 아니다.
- 멤버와 달리, 쓰이는 시점에 선언과 동시에 인스턴스가 만들어진다.
- 👉🏻 정적 문맥에서라도 상수 변수 이외의 정적 멤버는 가질 수 없다. 즉, 상수 표현을 위해 초기화된 final 기본 타입과 문자열 필드만 가질 수 있다. 

> 🚨 익명 클래스의 제약 사항
> - 선언한 지점에서만 인스턴스를 만들 수 있고, instatnceof 검사나 클래스의 이름이 필요한 작업은 수행할 수 없다.
> - 익명 클래스를 사용하는 클라이언트는 그 익명 클래스가 상위 타입에서 상속한 멤버 외에는 호출할 수 없다.
> - 여러 인터페이스를 구현할 수 없고, 인터페이스를 구현하는 동시에 다른 클래스를 상속할 수도 없다.
> - 익명 클래스를 사용하는 클라이언트는 그 익명 클래스가 상위 타입에서 상속한 멤버 외에는 호출할 수 없다.
> - 익명 클래스는 표현식이 중간에 등장하므로 짧게 작성 해야한다(10줄 이하로)

- 익명 클래스 예시
``` java
List<Integer> list = Arrays.asList(10, 5, 6, 7);
Collections.sort(list, new Comparator<Integer>() {
    @Override
    public int compare(Integer o1, Integer o2) {
        return Integer.compare(o1, o2);
    }
});

System.out.println(list);
```

(Java 8 이후 람다 등장)
```java
Collections.sort(list, (o1, o2) -> Integer.compare(o1, o2));
```

---

### 2-4. 지역 클래스
- 가장 드물게 사용된다.
- 지역 변수를 선언할 수 있는곳에서 선언 가능하고 유효 범위도 지역변수와 같다.
- 멤버 클래스처럼 이름이 있고, 반복해서 사용할 수 있다.
- 익명 클래스처럼 비정적 문맥에서 사용될 때만 바깥 인스턴스를 참조할 수 있다. 
- 정적 멤버는 가질 수 없으며 가독성을 위해 짧게 작성되어야 한다.

<details>
<summary>지역 클래스 예시</summary>
<div markdown="1">

```java
public class TestThread {
    public static void main(String[] args) {
        Thread goodNightThread = new Thread(() -> {
            try {
                for (int i = 0; i <10; i++){
                    Thread.sleep(1000);
                    System.out.println("양 " + i + "마리...");
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
        goodNightThread.start();
    }
}
```

</div>
</details>

- [Local inner class](https://velog.io/@mimah/Java-Local-inner-class) 의 또 다른 예제

# 3. 중첩 클래스의 사용 당위성 📚

- 메서드 밖에서도 사용해야 하거나 메서드 안에 정의하기엔 너무 길다면 멤버 클래스로 만든다.
- 멤버 클래스의 인스턴스 각각이 바깥 인스턴스를 참조한다면 비정적으로, 그렇지 않으면 정적으로 만들자.
- 중첩 클래스가 한 메서드 안에서만 쓰이면서 그 인스턴스를 생성하는 시점이 단 한곳이고 해당 타입으로 쓰기에 적합한 클래스나 인터페이스가 이미 있다면 익명 클래스로 만들고, 그렇지 않으면 지역 클래스로 만들자.

# 4. 회고 🧹

1. 멤버 클래스의 활용도에 대해 다시 생각 해 볼 수 있었음.
2. 메모리 구조에 대해 알고 있어야 해당 주제를 이해하기 용이함.
3. 책에 쓰여진 용어와 문장이 가독성이 좋지 못해 여러 레퍼런스를 참고 하려고 했지만 여러 레퍼런스 역시도 속 시원한 정리가 많지 않았음.
4. 해당 주제는 차후 정리가 더 필요한 분야라고 생각이 됨.

---
> - 자료 참조 및 출처
> - https://incheol-jung.gitbook.io/docs/q-and-a/java/static
> - https://refactoring.guru/ko/design-patterns/adapter
> - https://github.com/Meet-Coder-Study/book-effective-java/blob/main/4%EC%9E%A5/24_%EB%A9%A4%EB%B2%84%20%ED%81%B4%EB%9E%98%EC%8A%A4%EB%8A%94%20%EB%90%98%EB%8F%84%EB%A1%9D%20static%EC%9C%BC%EB%A1%9C%20%EB%A7%8C%EB%93%A4%EB%9D%BC_jiae.md
> - https://velog.io/@mimah/Java-Local-inner-class

**Comment by JoisFe** [comment, marked as answer]
#47 comment에 남겨놓았듯 단순히 멤버클래스 하나 만드는데도 메모리 낭비를 하지 않는지, 숨은 외부 참조를 가져 가비지 컬렉터가 활동하지 못하게 하는지 고민해야할 것이 많군요..

아 그리고 익명 클래스 부분 또한 유익했습니다.
자바 8 버전 이후 추상 메서드 1개 있는 인터페이스 즉 함수형 인터페이스를 구현하기 위해 람다를 이용하여 코드를 현저히 줄일 수 있다는 점이 좋았습니다.

1. 추상 메서드가 하나 있는 인터페이스를 상황에 맞게 매번 구현하기 위해 클래스를 만드는 점이 매우 불편 (특히 특정 한 부분에서만 해당 클래스 객체가 사용된다면 ?  --> 너무 비효율적)
2. 1 상황을 해결하기 위해 익명 클래스 개념이 나옴 하지만 이 또한 굳이 객체 생성 선언문, 메서드 명 등등 비효율적인 상황이 존재 (유일한 추상메서드 하나를 구현하기에 굳이 메서드 명 등을 입력할 이유가 없음)
3. 2 상황 마저 해결하기 위해 자바 8에서 람다가 등장

### 람다식은 메서드를 하나의 식으로 표현하는 선언적 프로그래밍 방법
- 저는 예전에 오해를 했습니다 람다는 메서드 자체인 줄
- 람다는 메서드 자체가 아니라 실행문을 가진 객체 !!

**Reply by Irisation23** [reply]
Arrow Function 을 기가막히게 지원하는 자바 스크립트가 근본 언어인건 확실한것 같네요 👍

---

![image](https://user-images.githubusercontent.com/53285909/211371856-98abbdfa-862c-4725-83d8-575a5cbd100b.png)

---

![image](https://user-images.githubusercontent.com/53285909/211372498-5053a7d5-a393-4df7-9bd1-8c18556380ac.png)

**Reply by JoisFe** [reply]
![image](https://user-images.githubusercontent.com/90208100/211374742-7f9ff39b-0f36-45b4-936d-96697d9665d4.png)
